runtime/src/mm/cpp/ObjectFactory.hpp

```diff
@@ -250,25 +251,74 @@ class ObjectFactoryStorage : private Pinned {
     SpinLock mutex_;
 };
 
+template <typename BaseAllocator, typename GC>
+class AllocatorWithGC {
+public:
+    AllocatorWithGC(BaseAllocator base, GC& gc) noexcept : base_(std::move(base)), gc_(gc) {}
+
+    void* Alloc(size_t size, size_t alignment) noexcept {
+        gc_.SafePointAllocation(size);
+        void* ptr = base_.Alloc(size, alignment);
+        if (ptr) {
+            return ptr;
+        }
+        gc_.OnOOM(size);
+        ptr = base_.Alloc(size, alignment);
+        if (ptr) {
+            return ptr;
+        }
+        konan::consoleErrorf("Out of memory trying to allocate %zu bytes. Aborting.\n", size);
+        konan::abort();
+    }
+
+private:
+    BaseAllocator base_;
+    GC& gc_;
+};
+
 } // namespace internal
 
 class ObjectFactory : private Pinned {
 public:
+    class Allocator {
+    public:
+        void* Alloc(size_t size, size_t alignment) noexcept { return konanAllocAlignedMemory(size, alignment); }
+    };
+
     using Storage = internal::ObjectFactoryStorage<kObjectAlignment>;
 
+    template <typename GC>
     class ThreadQueue : private MoveOnly {
     public:
-        explicit ThreadQueue(ObjectFactory& owner) noexcept : producer_(owner.storage_) {}
+        ThreadQueue(ObjectFactory& owner, GC& gc) noexcept : producer_(owner.storage_, internal::AllocatorWithGC(Allocator(), gc)) {}
+
+        ObjHeader* CreateObject(const TypeInfo* typeInfo) noexcept {
+            RuntimeAssert(!typeInfo->IsArray(), "Must not be an array");
+            size_t allocSize = typeInfo->instanceSize_;
+            auto& node = producer_.Insert(allocSize);
+            auto* object = static_cast<ObjHeader*>(node.Data());
+            object->typeInfoOrMeta_ = const_cast<TypeInfo*>(typeInfo);
+            return object;
+        }
 
-        ObjHeader* CreateObject(const TypeInfo* typeInfo) noexcept;
-        ArrayHeader* CreateArray(const TypeInfo* typeInfo, uint32_t count) noexcept;
+        ArrayHeader* CreateArray(const TypeInfo* typeInfo, uint32_t count) noexcept {
+            RuntimeAssert(typeInfo->IsArray(), "Must be an array");
+            uint32_t arraySize = static_cast<uint32_t>(-typeInfo->instanceSize_) * count;
```

Comment:
``-typeInfo->instanceSize_`` why negative?

Reply:
It's negative for the arrays: https://github.com/JetBrains/kotlin-native/blob/bb397ed8cc31f8f755ccaef93aa0f92ca61296d6/runtime/src/main/cpp/TypeInfo.h#L112